src/web/js/output-ui.js

```diff
@@ -20,58 +20,153 @@
-      }
-      marksByEditor[source].push({editor: editor, from: from, to: to, opts: opts, cb: cb});
-    }
-
-    function runMarks() {
-      var allMarks = marksByEditor;
-      var chunkSize = 100;
-      marksByEditor = {};
-      for (var source in allMarks) {
-        var marks = allMarks[source];
-        var editor = marks[0].editor; // these should be uniformly the same editor...
-        for (var i = 0; i < Math.ceil(marks.length / chunkSize); i++) {
-          function chunk(start, editor, marks) {
-            return function() {
-              editor.operation(function() {
-                for (var j = start; j < marks.length && j < start + chunkSize; j++) {
-                  //...but we'll use the right editor, just in case
-                  var mark = marks[j];
-                  var handle = mark.editor.markText(mark.from, mark.to, mark.opts);
-                  if (mark.cb)
-                    mark.cb(mark.editor, handle);
-                }
-              });
-            };
-          }
-          util.suspend(chunk(i * chunkSize, editor, marks));
+    var Position = function() {
+      
+      function cached_find(doc, positionCache, textMarker) {
+        var changeGeneration = doc.changeGeneration();
+        if (positionCache.has(changeGeneration))
+          return positionCache.get(changeGeneration);
+        else {
+          var pos = textMarker.find();
+          positionCache.set(changeGeneration, pos);
+          return pos;
         }
       }
-    }
-
-
-    function mapK(inList, f, k, outList) {
-      if (inList.length === 0) { k(outList || []); }
-      else {
-        var newInList = inList.slice(1, inList.length);
-        f(inList[0], function(v) {
-          mapK(newInList, f, k, (outList || []).concat([v]))
+    
+      function Position(doc, source, from, to, inclusiveLeft, inclusiveRight) {
+        var inclusiveLeft   = false,
+            inclusiveRight  = false;
+        if (inclusiveLeft === undefined)
+          inclusiveLeft = false;
+        if (inclusiveRight === undefined)
+          inclusiveRight = false;
+        this.inclusiveLeft  = inclusiveLeft;
+        this.inclusiveRight = inclusiveLeft;
+        
+        this.doc = doc;
+        this.source = source;
+        
+        var textMarker = doc.markText(from, to, this.options);
+        this._textMarker = textMarker;
+          
+        var positionCache = new Map();
+          
+        Object.defineProperty(this, 'from', {
+          get: function() {
+            var pos = cached_find(doc, positionCache, textMarker);
+            return pos !== undefined ? pos.from : undefined;
+          }
+        });
+        
+        Object.defineProperty(this, 'to', {
+          get: function() {
+            var pos = cached_find(doc, positionCache, textMarker);
+            return pos !== undefined ? pos.to : undefined;
+          }
         });
+        
+        positionCache.set(doc.changeGeneration(), {from, to});
```

Comment:
@jpolitz Addressed by 60517497cca81e2413fc236eeb3160a0412e00f9

This was purely a mistake. TIL about object shorthand.